tensorbay/client/profile.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+#
+# Copyright 2021 Graviti. Licensed under MIT License.
+#
+
+"""Statistical.
+
+:class:`Statistical` is a class used to statistical speed of interface.
+:class:`Config` is a class defining the concept of Request Config.
+
+"""
+
+import json
+import time
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Callable, DefaultDict, TypeVar
+
+from requests.models import Response
+from requests_toolbelt import MultipartEncoder
+
+_Callable = TypeVar("_Callable", bound=Callable[..., Response])
+SUMMARY: DefaultDict[str, DefaultDict[str, Any]] = defaultdict(lambda: defaultdict(int))
+
+
+def _synchroniza_info(key: str, response: Response, cost_time: float, data_length: int) -> None:
+    SUMMARY[key]["totalTime"] += cost_time
+    SUMMARY[key]["callNumber"] += 1
+    SUMMARY[key]["totalResponseLength"] += len(response.content)
+    SUMMARY[key]["totalFileSize"] += data_length
+
+
+class Config:
+    """This is a base class defining the concept of Request Config.
+
+    Attributes:
+        max_retries: Maximum retry times of the request.
+        allowed_retry_methods: The allowed methods for retrying request.
+        allowed_retry_status: The allowed status for retrying request.
+            If both methods and status are fitted, the retrying strategy will work.
+        timeout: Timeout value of the request in seconds.
+        is_internal: Whether the request is from internal.
+        start_statistical: Whether to start statistical recording.
+
+    """
+
+    def __init__(self) -> None:
+
+        self.max_retries = 3
+        self.allowed_retry_methods = ["HEAD", "OPTIONS", "POST", "PUT"]
+        self.allowed_retry_status = [429, 500, 502, 503, 504]
+
+        self.timeout = 30
+        self.is_internal = False
+        self._x_source = "PYTHON-SDK"
+        self.start_statistical = False
+
+
+config = Config()
+
+
+def statistical(func: _Callable) -> _Callable:
+    """A decorator to record the call times, running time, response length ..etc of the function.
+
+    Arguments:
+        func: The needed statistical function.
+
+    Returns:
+        The wrapped function.
+
+    """
+
+    @wraps(func)
+    def wrapper(*arg: Any, **kwargs: Any) -> Response:
+        if not config.start_statistical:
+            return func(*arg, **kwargs)
+
+        key = f"[{arg[1]}] {arg[2]}"
+        start_time = time.time()
+        data: MultipartEncoder = kwargs.get("data", {})
+        data_length = data.len if data else 0
+        response = func(*arg, **kwargs)
+        cost_time = time.time() - start_time
+        _synchroniza_info(key, response, cost_time, data_length)
+
+        return response
+
+    return wrapper  # type: ignore[return-value]
+
+
+def save_statistical(path: str) -> None:
```

Comment:
the statistical feature needs a start and an end, 
can we design it as a `with` block?